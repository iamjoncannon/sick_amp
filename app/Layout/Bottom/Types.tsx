export interface SongObject {
    ID: number 
    ALBUM: string 
    TITLE: string 
    ARTIST: string 
    TRACKNUMBER: string
}

export interface songStateType {
    [ key: number ] : SongObject
}

export interface Playlist {
    Title: string
    id: number
    ids: number[]
}

export interface Store {
    SelectedPlaylist: number
    PlayLists : Playlist[] | void 
    Songs: songStateType | void 
    ColumnHash : { [ Column : string] : Boolean} | void 
}

