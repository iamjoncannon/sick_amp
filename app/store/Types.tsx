export interface SongObject {
    ID: number 
    ALBUM: string 
    TITLE: string 
    ARTIST: string 
    TRACKNUMBER: string
    FILENAME: string
}

export interface songStateType {
    [ key: number ] : SongObject
}

export interface Playlist {
    Title: string
    id: number
    ids: number[]
}

export interface PlaylistState {
    [key : number] : Playlist
}

export interface Transport {
    previous: number | void 
    current: number | void
    next: number | void
}

export interface Store {
    Transport : Transport 
    isPlaying: Boolean
    SelectedPlaylist: number | string | void
    RunningPlaylist: number | string | void
    PlayLists : PlaylistState | void 
    Songs: songStateType | void 
    ColumnHash : { [ Column : string] : Boolean} | void 
    draggedOverPlaylist: Boolean    
}

