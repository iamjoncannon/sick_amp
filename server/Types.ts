export interface SongObject {
    ID: number 
    ALBUM: string 
    TITLE: string 
    ARTIST: string 
    TRACKNUMBER: string
    FILENAME? : string
}

export interface songStateType {
    [ key: number ] : SongObject
}

export interface Playlist {
    Title: string
    id: number
    ids: number[]
}

export interface returnDataType {
    Songs: songStateType
    PlayLists: Playlist[]
}
