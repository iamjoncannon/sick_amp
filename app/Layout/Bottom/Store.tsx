import React from 'react';
import * as Types from './Types'

const initialState : Types.Store = {
    isHydrated : false,
    SelectedPlaylist: 1,
    PlayLists: null,
    Songs: null,
};

interface ReduxAction {
    type: string
    payload: string
}

function reducer(state : Types.Store, action : ReduxAction ) {

    switch (action.type) {

        case 'HYDRATE': {

            return {...state, Songs: action.payload, isHydrated : true }
        }

        case 'SELECT_PLAYLIST':
        
            return { ...state, SelectedPlaylist: action.payload }
        
        case 'ADD_SONG_TO_PLAYLIST':

            return {...state}
        
        default:
          return state;
      }
}

export const Store = React.createContext(null);

export function StoreProvider(props : any) {

    const [state, dispatch] = React.useReducer(reducer, initialState);
    const value = { state, dispatch };

    return (
        <Store.Provider value={value}>
            {props.children}
        </Store.Provider>
    )
}
