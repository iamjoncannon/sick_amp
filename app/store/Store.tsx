import React from 'react';
import * as Types from './Types'

const initialState : Types.Store = {
    Transport : {
        current: 1, 
    },
    isPlaying: false,
    SelectedPlaylist: "All",
    RunningPlaylist: "All",
    Columns: null,
    PlayLists: null,
    Songs: null,
    draggedOverPlaylist: null,
    token: "dev"
};

interface ReduxAction {
    type: string
    // ideally this would be an enum 
    // with the contract for every action
    payload: any 
}

function reducer(state : Types.Store, action : ReduxAction ) {

    console.log("Action: ", action.type, action.payload)

    switch (action.type) {

        case 'HYDRATE_PLAYLISTS': {

            return {...state, PlayLists: action.payload}
        }

        case 'HYDRATE_SONGS': {

            const { PlayList, Page, data } = action.payload

            // pouring songs into the pool

            const Songs = {...state.Songs, ...data}

            // update the hydration hash for the playlist 

            const PlayLists = {...state.PlayLists}

            PlayLists[PlayList].hydrated[Page] = true 
  
            return {...state, PlayLists : PlayLists, Songs: Songs }
        }

        case 'PLAYLIST_END': {

            const { PlayList, Page } = action.payload

            const next_PlayLists = {...state.PlayLists}

            next_PlayLists[PlayList].hydrated["Complete"] = true 
            next_PlayLists[PlayList].hydrated[Page] = true 

            return {...state, PlayLists: next_PlayLists }
        }

        case 'HYDRATE_COLUMNS': {

            return {...state, Columns: action.payload.data}
        }
        

        case 'SELECT_PLAYLIST':
        
            return { ...state, SelectedPlaylist: action.payload }

        case 'PLAY_TRACK': {

            const current : number = Number(action.payload)

            const { SelectedPlaylist } = state 

            let newTransport : Types.Transport = { current } 

            return {...state, Transport: newTransport, isPlaying: true, RunningPlaylist: SelectedPlaylist}
        }
        
        case "PLAY_PREVIOUS_TRACK": {

            const { RunningPlaylist, PlayLists, Transport : { current } } = state 

            // architecture note- I decided to structure the "All" Playlist and the specific
            // playlists differently- see ./Types -- this requires us to handle each differently
            // in the playlist actions - this could certainly be revisited and refactored 

            // if its playing from the all playlist, then its simply the previous track
            // or the end of the playlist if current = 0 

            let next_current 

            if(RunningPlaylist === "All"){

                next_current = current === 0 ? PlayLists[RunningPlaylist].length -1 : Number(current) - 1 
            }
            else{

                // if its a specific playlist, then we need to find the index of the track in the 
                // playlists ids and return the previous index, or end if 0 

                let CurrentPlaylist = PlayLists[RunningPlaylist].ids

                const current_index = CurrentPlaylist.indexOf(current)

                next_current = current_index === 0 ? CurrentPlaylist[CurrentPlaylist.length -1] : CurrentPlaylist[current_index -1 ] 
            
            }

            let newTransport : Types.Transport = { current: next_current } 
            
            return {...state, Transport: newTransport, isPlaying: true}
        }

        case "PLAY_NEXT_TRACK": {

            const { RunningPlaylist, PlayLists, Transport : { current } } = state 

            let next_current 
            let CurrentPlaylist 

            if(RunningPlaylist === "All"){

                CurrentPlaylist = PlayLists[RunningPlaylist]

                next_current = current === CurrentPlaylist.length -1 ? 0 : Number(current) + 1 
            }
            else{

                CurrentPlaylist = PlayLists[RunningPlaylist].ids

                const current_index = CurrentPlaylist.indexOf(current)

                next_current = current_index === CurrentPlaylist.length -1 ? CurrentPlaylist[0] : CurrentPlaylist[current_index + 1 ] 
            }

            let newTransport : Types.Transport = { current: next_current } 
            
            return {...state, Transport: newTransport, isPlaying: true}
        }

        case 'ADD_SONG_TO_PLAYLIST':{

            const { PlayLists } = state 
            const { payload: { song, playlist } } = action

            const nextPlaylists = [...PlayLists]

            if(!nextPlaylists[playlist].ids.includes(Number(song))){

                nextPlaylists[playlist].ids = [...PlayLists[playlist].ids, Number(song)]
            }

            return {...state, PlayLists : nextPlaylists, draggedOverPlaylist: null}
        }

        case "TOGGLE_PLAYERSTATE": {

            return {...state, isPlaying: !state.isPlaying}
        }


        case "REARRANGE_PLAYLIST": {

            const { item_to_put_before, item_to_be_moved } = action.payload

            let next_playlist_object = [...state.PlayLists]
            
            if(state.SelectedPlaylist !== "All"){

                let targetPlaylist = next_playlist_object[state.SelectedPlaylist].ids

                let nextTargetPlaylist = []

                // this is not the most clever way to do this
                // but it works 
                for(let i = 0; i < targetPlaylist.length; i++){

                    if(targetPlaylist[i] !== item_to_be_moved){
                        
                        if(targetPlaylist[i] === item_to_put_before){

                            nextTargetPlaylist.push(item_to_be_moved)
                        }

                        nextTargetPlaylist.push(targetPlaylist[i])
                    }
                }

                next_playlist_object[state.SelectedPlaylist].ids = nextTargetPlaylist
                
            }
            
            // this could be refactored- playlists is an object
            // we also want to treat like an array - "All" needs
            // to be reappended if we destructure as above 
                
            next_playlist_object["All"] = [...Object.values(state.Songs)]
                        
            return {...state, PlayLists : next_playlist_object}
        }

        case "ADD_PLAYLIST":{

            let next_playlist_object = {...state.PlayLists}

            next_playlist_object[next_playlist_object.length] = { name: "New PlayList", id: next_playlist_object.length, files: [ ]}

            return {...state, PlayLists : next_playlist_object}
        }

        case "DRAG_OVER_PLAYLIST": {

            return {...state, draggedOverPlaylist: action.payload}
        }
        
        default:
          return state;
      }
}

export const Store = React.createContext(null);

export function StoreProvider(props : any) {

    const [state, dispatch] = React.useReducer(reducer, initialState);
    const value = { state, dispatch };

    return (
        <Store.Provider value={value}>
            {props.children}
        </Store.Provider>
    )
}

function clone(item){

    return JSON.parse(JSON.stringify(item))
}