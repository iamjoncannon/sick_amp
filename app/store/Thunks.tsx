const { default: axios } = require('axios');
import { api_url } from '../../config'

// called by PlayLists 

export async function fetchInitialData(token: string, dispatch: any){

    axios.get(`${api_url}/folders/?api_key=${token}`).then(( { data } )=>{
    
        const formattedData = {}
        
        for(let playlist in data){

            formattedData["All"] = { name: "All Songs", hydrated: {}  }
            
            // make sure the local hash key is the same 
            // as the id returned from the server
            formattedData[data[playlist].id] = data[playlist]
            
            // add hydrated boolean to prevent manage
            // fetching of data for each playlist 
            formattedData[data[playlist].id].hydrated = {}
        }

        dispatch({
            type: "HYDRATE_PLAYLISTS",
            payload: formattedData
        })
        
    }).catch(err=>console.log(err))

    hydrateSongs(token, dispatch, "All", 1)
}

// called by EditablePlayList

export async function postPlayList(name: string, token: string, dispatch: any){

    let result

    try{

        result = await axios.post(`${api_url}/folders/?api_key=${token}`, { name })
    }
    catch(err){

        console.log(err)
        return 
    }

    dispatch({type: 'UPDATE_PLAYLISTS', payload: result.data})
}

export async function putPlayList(name: string, token: string, dispatch: any){

    const endpoint = `${api_url}/folders/?api_key=${token}`
    let result
    
    try{
        
        result = await axios.put(endpoint, { name })
    }
    catch(err){

        console.log(err)
        dispatch({type: "CANCEL_UPDATE_PLAYLISTS"})
        return 
    }
    
    dispatch({type: 'UPDATE_PLAYLISTS', payload: result.data})
}

// addSongToPlaylist POST /files/{file_id}/folders/{folder_id}

export async function addSongToPlaylist(folder_id: number, song_id: number, token: string, dispatch: any){

    const endpoint = `${api_url}/files/${song_id}/folders/${folder_id}?api_key=${token}`

    let result 

    try {

        result = await axios.post(endpoint)
    }
    catch(err){

        console.log("addSongToPlaylist thunk error: ", err)
        return
    }

    console.log(result)


    /*

    dispatch({
                type: "ADD_SONG_TO_PLAYLIST",
                payload: {
                    newPlayListObject = result.data,
                    playlist_id: folder_id
                }
            })

    */

}


export async function rearrange_playlist(folder_id: number, song_id: number, position: number, token: string, dispatch: any){

    // files/{fileid}/folders/{folderid} { position: int }
    const endpoint = `${api_url}/files/${song_id}/folders/${folder_id}?api_key=${token}`
    
    let result
    
    try{
        
        result = await axios.put(endpoint, { position })
    }
    catch(err){

        console.log(err)
        dispatch({type: "CANCEL_UPDATE_PLAYLISTS"})
        return 
    }

    dispatch({
        type: "REARRANGE_PLAYLIST",
        payload: {
            
            updatedPlaylist: result.data
        }
    })

}


// called by Songs 

export async function hydrateColumns(token: string, dispatch: any){

    let data 

    try{
        data = await axios.get(`${api_url}/fields?api_key=${token}`)
    }
    catch(err){
        console.log(err)
    }

    dispatch({
        type: "HYDRATE_COLUMNS",
        payload: data
    })

}

// called by SongTable 

export async function hydrateSongs(token: string, dispatch: any, PlayList: string, Page: number){

    axios.get(`${api_url}/files/?page=${Page}&api_key=${token}`).then(( { data } )=>{
    
        const formattedData = {}

        for(let song in data){

            const { id } = data[song]

            formattedData[id] = data[song]
        }

        if(Object.keys(formattedData).length === 0){

            dispatch({
                type: "PLAYLIST_END",
                payload: { PlayList, Page }
            })
        }
        else{
            
            dispatch({
                type: "HYDRATE_SONGS",
                payload: { PlayList: PlayList, Page, data: formattedData }
            })
        }

    }).catch(err=>console.log(err))
}



