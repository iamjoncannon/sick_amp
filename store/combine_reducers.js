import { combineReducers } from 'redux'
import User_state from './User/reducer_for_User'

export default combineReducers({
 User_state,
})