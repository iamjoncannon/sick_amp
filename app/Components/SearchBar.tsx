import React from 'react';
import styled from 'styled-components'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSearch } from '@fortawesome/free-solid-svg-icons'

import { Store } from '../store/Store'

interface SectionContainerProps {
    all_fields : string | boolean
}

const SectionContainer = styled.div<SectionContainerProps>`

    display: flex;
    flex-direction: column;

    svg {
        color: ${props=>props.theme.fontColor};
        position: relative;
        bottom: ${props=> props.all_fields ? "3vh" : "2vh"}; 
        margin-bottom: -1.4vh;
        height: ${props=> props.all_fields ? "2vh" : "1.5vh"};

    }

    input {
        background-image: ${props=>props.theme.secondaryColor_Background};
        color: ${props=>props.theme.fontColor};
        border: none;
        font-size: ${props=> props.all_fields ? "3vh" : "1.6vh"}; 
    }
`

interface SearchBarProps {

    target: string
}

const SearchBar = (props : SearchBarProps) => {

    const whichStore = process.env.NODE_ENV === 'test' ? require('./SearchBar.test').MockStore : Store ;

    const { state, dispatch } = React.useContext(whichStore);

    const { SearchBarText } = state

    const  all_fields  = SearchBarText.all_fields.text

    const { target } = props 

    function handleChange(e){

        dispatch({type: "HANDLE_SEARCHBAR_TEXT", payload: { target: target, text: e.target.value} })
    }

    const toggleSearchBar = (state_change) => {

        dispatch({type: "TOGGLE_SEARCHBAR_FOCUS", payload: state_change})
    }

    const inlines = props.target === "all_fields" ? { width: "20vw", height: "4vh"} : { width: "10vw", height: "2vh"} 

    return (

        <SectionContainer all_fields={props.target === "all_fields"}>
            
            <input 
                style={ inlines }
                onChange={handleChange}
                value={ all_fields ? all_fields : SearchBarText[target].text }   
                onFocus={()=>toggleSearchBar(true)} 
                onBlur={()=>toggleSearchBar(false)}    
            />
                { SearchBarText[target].text === "" && 
                  all_fields === "" && 

                    <FontAwesomeIcon 
                        icon={faSearch} 
                    />
                }
        
        </SectionContainer>
    )
}

export default SearchBar