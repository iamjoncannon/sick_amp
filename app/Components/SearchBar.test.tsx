import React from 'react'
import { mount } from 'enzyme'
import SearchBar from './SearchBar'
import Types from '../store/Types'
import {TOGGLE_SEARCHBAR_FOCUS, HANDLE_SEARCHBAR_TEXT} from '../store/Store'

/*

    Intended behavior: 

    [x] renders an icon if the text is empty

    [x] onFocus and blur, dispatches a toggle search bar to global state 

    [x] appends its value to global state at SearchBarText[target].text

    [x] inline styling changes based on which field is getting rendered

*/

export let MockStore 
let mockState 
let dispatchedActions = []
let mockInitialState

function mockReducer(state : Types.Store, action : ReduxAction ) {
    
    switch (action.type) {

        case "TOGGLE_SEARCHBAR_FOCUS": {

            return TOGGLE_SEARCHBAR_FOCUS(state, action)
        }

        case "HANDLE_SEARCHBAR_TEXT" : {

            return HANDLE_SEARCHBAR_TEXT(state, action)
        }
    }
}

describe("SearchBar", ()=>{

    beforeEach(() => {

        MockStore = React.createContext(null);

        mockInitialState = {

            isTypingInSearchBar: true,
            SearchBarText: {
              '1': { field: 'artist', text: '' },
              '2': { field: 'album_artist', text: '' },
              '3': { field: 'genre', text: '' },
              all_fields: { text: '' }
            }
        }
    });

    function MockStoreProvider(props : any) {

        let [ state, dispatch ] = React.useReducer(mockReducer, mockInitialState);

        mockState = state 

        let mockDispatch = function(...args){

            dispatchedActions.push(...args)

            dispatch(...args)
        }

        return (
            <MockStore.Provider value={{ state, dispatch: mockDispatch }}>
                {props.children}
            </MockStore.Provider>
        )
    }

    function TestComponent (prop){

        return mount(
            <MockStoreProvider>
                <SearchBar target={prop}/>
            </MockStoreProvider>
            );
    }

    it('SearchBar renders correctly with a target prop', () => {

        const Tree = TestComponent("all_fields")
        
        expect(Tree).toMatchSnapshot();
    });

    it('onFocus and blur dispatches TOGGLE_SEARCHBAR_FOCUS action to global state', ()=>{

        const Tree = TestComponent("all_fields")
        
        const input = Tree.find('input')

        input.simulate("focus")

        expect(dispatchedActions[0].type).toEqual("TOGGLE_SEARCHBAR_FOCUS")
        expect(dispatchedActions[0].payload).toEqual(true)
    
        input.simulate("blur")

        expect(dispatchedActions[1].type).toEqual("TOGGLE_SEARCHBAR_FOCUS")
        expect(dispatchedActions[1].payload).toEqual(false)
    })

    it('SearchBar only renders the svg icon if the text field is empty', ()=>{

        const Tree = TestComponent("all_fields")
        
        expect(Tree.find("svg").length).toEqual(1)

        mockInitialState.SearchBarText.all_fields.text = "This is test input"

        const Tree2 = TestComponent("all_fields")
            
        expect(Tree2.find("svg").length).toEqual(0)
    })

    it("onChange callback calls HANDLE_SEARCHBAR_TEXT and appends its value to global state at SearchBarText[target].text", ()=>{

        dispatchedActions = []

        const Tree = TestComponent("all_fields")
        
        let input = Tree.find('input')

        const testText = "This is my test text"

        input.simulate("change", {target: {value: testText}})

        expect(mockInitialState.SearchBarText.all_fields.text).toEqual(testText)
        expect(dispatchedActions[0].type).toEqual("HANDLE_SEARCHBAR_TEXT")
        expect(dispatchedActions[0].payload.text).toEqual(testText)

        const Tree2 = TestComponent('1')
        
        input = Tree2.find('input')

        input.simulate("change", {target: {value: testText}})

        expect(mockInitialState.SearchBarText['1'].text).toEqual(testText)
        expect(dispatchedActions[1].type).toEqual("HANDLE_SEARCHBAR_TEXT")
        expect(dispatchedActions[1].payload.text).toEqual(testText)
    })

    it("inline styling changes based on which field is selected", ()=>{

        const Tree = TestComponent("all_fields")

        let input = Tree.find('input')

        expect(input.props().style.width).toEqual('20vw')
        expect(input.props().style.height).toEqual('4vh')

        const Tree2 = TestComponent('1')
        
        input = Tree2.find('input')

        expect(input.props().style.width).toEqual('10vw')
        expect(input.props().style.height).toEqual('2vh')
    })

})
