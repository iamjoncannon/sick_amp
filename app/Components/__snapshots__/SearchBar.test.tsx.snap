// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar SearchBar renders correctly with a target prop 1`] = `ReactWrapper {}`;
